flight/src/types.rs

```diff
@@ -60,9 +62,19 @@ impl TryFrom<unvalidated::Event<Ipld>> for ConclusionInit {
     fn try_from(event: unvalidated::Event<Ipld>) -> Result<Self> {
         let init_payload = event
             .init_payload()
-            .ok_or_else(|| anyhow!("no init payload found"))?;
+            .ok_or_else(|| anyhow!("malformed event: no init payload found"))?;
+        let model = init_payload.header().model();
+        let stream_type = StreamId::try_from(model)
```

Comment:
Do we want to move this logic to `init.rs` or should it live here? I prefer it to be here since this seems like an inference of what we pass into the struct. The init header struct does not add any checks while populating this field. Any thoughts on this?

Reply:
I like the logic here as its part of the mapping from raw event to conclusion event. However this logic is incorrect and only passes tests because the tests are also incorrect.

This is subtle, let me start with the structure of an stream id.

As you have seen a stream id is an encoded tuple of (stream type, init cid).
There are two stream ids at play in an init event. First, the stream id of the init event itself, lets call this one the init_stream_id. Second the stream id of the model that the stream belongs to, lets call this one the model_stream_id.

Therefore a given init event the model_stream_id is found in the model field of the header. This is the one the code above is decoding. The init_stream_id is not part of the init event since it is the cid of the init event itself. 

The stream id we need to construct for a conclucsion event is the init_stream_id, i.e. the stream id of the stream to which the event belongs. 

The init_stream_id can have two possible stream types, Model or ModelInstanceDocument depending. However the model_stream_id is always a stream type of Model. This is because a stream cannot belong to a ModelInstanceDocument stream. 

So back to the code above. The code is decoding the model_stream_id type and using that as the stream type for the init_stream_id which is incorrect. However if you look into the [init_event](https://github.com/ceramicnetwork/rust-ceramic/blob/9c2575963f421bf8052966aafc577403dd649e25/event-svc/src/tests/mod.rs#L144) function is takes a stream id and uses it as the model field for the init event. However it does not guarantee that the stream type is always a Model type. If you add these lines to that function the test fails.

```rust
    assert_eq!(
        model.r#type,
        StreamIdType::Model,
        "an init event's model stream id must always be a model stream type"
    );
```

This stems from the `create_deterministic_stream_id_model` and `create_deterministic_stream_id_document` functions  and how they are used. The test creates two stream ids. Both of these stream ids are used as the model field in the init event. Therefore both of these stream ids should be model stream ids. This would make the above assertion pass but then the `ConclucsionInit::try_from` code would then always produce a stream type of model for the init_stream_id, which is incorrect. 

So how can we determine the type of the init_stream_id? This needs one more piece of information. Since model streams must also belong to a model stream there is a special known stream (sometimes called the metamodel) to which all model streams belong. We can use this information to recover the type of the init_stream_id. If the model of the init event is equal to the metamodel stream id then we known that the init_stream_id type is model. If its any other stream id then its a document type. 

All of this is to say the code above should do a simple equality comparison of the init event model field with the metamodel stream id to determine the stream type.

The metamodel stream id is `kh4q0ozorrgaq2mezktnrmdwleo1d` we should create a constant of this value along side the StreamId type. It comes from [here](https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/stream-model/src/model.ts#L121)

Reply:
As an aside we should probably create an MID builder that expects an actual stream id for the model value and can verify that it is a model stream instead of using the vanilla event builder. Not something we should do in this PR, but would be a good API design to catch this bug at compile time.

Reply:
Note the metamodel streamid is of type UNLOADABLE, not Model